packages/components/src/DateTimePickers/DateTimePicker/DateTimePicker.component.js

```diff
@@ -0,0 +1,160 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import getMonth from 'date-fns/get_month';
+import getYear from 'date-fns/get_year';
+
+import theme from './DateTimePicker.scss';
+import DateTimeView from '../views/DateTimeView';
+import MonthYearView from '../views/MonthYearView';
+
+class DateTimePicker extends React.Component {
+
+	constructor(props) {
+		super(props);
+
+		// eslint-disable-next-line
+		console.warn(`UNSTABLE WARNING: The 'DateTimePicker' and all the sub components aren't ready to be used in Apps. Code can (will) change outside the release process until it's ready.`);
+
+		const selectedDate = props.selection && props.selection.date;
+		const selectedTime = props.selection && props.selection.time;
+
+		const initialCalendarDate = selectedDate === undefined
+			? new Date()
+			: selectedDate;
+
+		this.state = {
+			isDateTimeView: true,
+			calendar: {
+				monthIndex: getMonth(initialCalendarDate),
+				year: getYear(initialCalendarDate),
+			},
+			selectedDate,
+			selectedTime,
+		};
+
+		this.setDateTimeView = this.setView.bind(this, true);
+		this.setMonthYearView = this.setView.bind(this, false);
+		this.onSelectCalendarMonth = this.onSelectCalendarMonth.bind(this);
+		this.onSelectCalendarYear = this.onSelectCalendarYear.bind(this);
+		this.onSelectCalendarMonthYear = this.onSelectCalendarMonthYear.bind(this);
+		this.onSelectDate = this.onSelectDate.bind(this);
+		this.onSelectTime = this.onSelectTime.bind(this);
+	}
+
+	componentWillReceiveProps(nextProps) {
+		const propsUpdated = this.props.selection !== nextProps.selection;
+
+		if (!propsUpdated) {
+			return;
+		}
+
+		const newSelectedDate = nextProps.selection && nextProps.selection.date;
+		const newSelectedTime = nextProps.selection && nextProps.selection.time;
+		const needToUpdateDate = newSelectedDate !== this.state.selectedDate;
+		const needToUpdateTime = newSelectedTime !== this.state.selectedTime;
+		const needToUpdateState = needToUpdateDate || needToUpdateTime;
+
+		if (!needToUpdateState) {
+			return;
+		}
+
+		const needToUpdateCalendar = needToUpdateDate && newSelectedDate !== undefined;
+
+		const newState = {
+			...(needToUpdateDate && { selectedDate: newSelectedDate }),
+			...(needToUpdateTime && { selectedTime: newSelectedTime }),
+			...(needToUpdateCalendar && {
+				calendar: {
+					monthIndex: getMonth(newSelectedDate),
+					year: getYear(newSelectedDate),
+				},
+			}),
+		};
+
+		this.setState(newState);
+	}
+
+	onSelectDate(selectedDate) {
+		this.setState({ selectedDate }, () => {
+			this.trySubmit();
+		});
```

Comment:
I prefer keep it separate : the handler and the action(s). Even if there is only one action which didn't take any args, it could change in the future. And when it could, it will.
Furthermore, the callback is called with two args that will be spread to the action which don't want it.
Implicit spreading args like that all over the place is hard to reason about.
I don't think reduce the call-stack of 1 and removing some couple of parentheses and curly brackets is more important than explicit function calls and args definition.

Reply:
✘